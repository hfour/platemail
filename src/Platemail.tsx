import * as React from 'react';
import * as ReactDOM from 'react-dom/server';
import PlatemailContext from './PlatemailContext';
import { RenderedEmailObject } from './types';

export default class Platemail {
  doctype: string = `<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">`;

  //TODO: Constructor - uses context, custom and existing components

  render(EmailTemplate: React.FunctionComponent<any>, props?: any): RenderedEmailObject {
    let html = (
      <PlatemailContext.Provider value={true}>
        <EmailTemplate {...props} />
      </PlatemailContext.Provider>
    );

    let text = (
      <PlatemailContext.Provider value={false}>
        <EmailTemplate {...props} />
      </PlatemailContext.Provider>
    );
    
    return { subject: html.props.children.props.subject, body: this.renderedBody(html, text) };
  }

  renderedBody(html: any, text: any) {
    return {
      html: [this.doctype, ReactDOM.renderToStaticMarkup(html as any)].join('\n'),
      text: ReactDOM.renderToStaticMarkup(text as any),
    };
  }

}
