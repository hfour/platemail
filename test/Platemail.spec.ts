import Platemail from "../src/Platemail";
import {
  simpleEmailTemplate,
  customEmailTemplate
} from "../examples/Templates";

describe("emails", () => {
  it("should match the existing snapshot", () => {
    const pm = new Platemail();
    let emailProps = { subject: "Simple Email Template" };
    let renderedEmailObject = pm.render(simpleEmailTemplate, emailProps);
    return expect(renderedEmailObject).toMatchSnapshot();
  });

  it("should allow custom components", () => {
    const pm = new Platemail();
    let customEmailProps = {
      subject: "Custom Email Template",
      user: { name: "Spike" }
    };
    let customEmailObject = pm.render(customEmailTemplate, customEmailProps);
    return expect(customEmailObject).toMatchSnapshot();
  });
});
