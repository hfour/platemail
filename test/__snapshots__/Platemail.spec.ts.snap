// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emails should allow custom components 1`] = `
Object {
  "body": Object {
    "html": "<!DOCTYPE html PUBLIC \\"-//W3C//DTD XHTML 1.0 Transitional//EN\\"
\\"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\\">
<html lang=\\"en\\"><head><meta http-equiv=\\"Content-Type\\" content=\\"text/html; charset=UTF-8\\"/><meta name=\\"format-detection\\" content=\\"telephone=no\\"/><meta name=\\"viewport\\" content=\\"width=device-width; initial-scale=1.0; maximum-scale=1.0; user-scalable=no;\\"/><meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=9; IE=8; IE=7; IE=EDGE\\"/><link href=\\"https://fonts.googleapis.com/css?family=Open+Sans:300,400,400i,600,600i,700,700i\\" rel=\\"stylesheet\\"/><style type=\\"text/css\\"></style></head><body><table><tr><td>Custom Email Template<p style=\\"font-size:20px\\">Welcome Spike!</p></td></tr></table></body></html>",
    "text": "Custom Email Template

Welcome Spike!

",
  },
  "subject": "Custom Email Template",
}
`;

exports[`emails should match the existing snapshot 1`] = `
Object {
  "body": Object {
    "html": "<!DOCTYPE html PUBLIC \\"-//W3C//DTD XHTML 1.0 Transitional//EN\\"
\\"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\\">
<html lang=\\"en\\"><head><meta http-equiv=\\"Content-Type\\" content=\\"text/html; charset=UTF-8\\"/><meta name=\\"format-detection\\" content=\\"telephone=no\\"/><meta name=\\"viewport\\" content=\\"width=device-width; initial-scale=1.0; maximum-scale=1.0; user-scalable=no;\\"/><meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=9; IE=8; IE=7; IE=EDGE\\"/><link href=\\"https://fonts.googleapis.com/css?family=Open+Sans:300,400,400i,600,600i,700,700i\\" rel=\\"stylesheet\\"/><style type=\\"text/css\\"></style></head><body><table><tr><td>Simple Email Template<p style=\\"font-size:20px\\">This is a simple template.</p></td></tr></table></body></html>",
    "text": "Simple Email Template

This is a simple template.

",
  },
  "subject": "Simple Email Template",
}
`;
